compiler/sem/mirexec.nim

```diff
@@ -142,93 +141,106 @@ func computeCfg*(tree: MirTree): ControlFlowGraph =
   ## 1. a control-flow graph needs to materialize all edges for a given `tree`
   ## 2. the amount of allocations/bookkeeping necessary to do so
   ##
-  ## The most amount of bookkeeping is required for finalizers and exception.
-
+  ## The most amount of bookkeeping is required for finalizers and exceptions.
   type
-    FinalizerState = object
-      span: Slice[NodePosition]
-      next: seq[NodePosition] ## the exits registered to the finalizer
-
     ClosureEnv = object
-      ## Using a real closure (i.e. ``.closure``) would be a bit too costly,
-      ## so a manual implementation is used instead
       instrs: seq[Instr]
-      finalizers: seq[FinalizerState]
-      joins: Table[NodePosition, JoinId]
-
-    JoinPoint = object
-      id: JoinId
-      node: NodePosition
-      isNew: bool
-
-  func considerFinalizer(target: NodePosition,
-                         f: var FinalizerState): NodePosition =
-    ## Redirects a jump to `target` through a finalizer if one is active and
-    ## the jump crosses its boundary
-    if target in f.span:
-      target
-    else:
-      if target notin f.next:
-        f.next.incl target
-
-      f.span.b
-
-  func addJoin(env: var ClosureEnv, dest: NodePosition): JoinPoint =
-    let
-      nextId = env.joins.len.JoinId
-      id = env.joins.mgetOrPut(dest, nextId)
-
-    JoinPoint(id: id, node: dest, isNew: id == nextId)
-
-  func addEdge(env: var ClosureEnv, a, b: NodePosition,
-               isLoop: bool): JoinPoint =
-    let target =
-      if isLoop:
-        assert a > b, "a loop CFG edge must describe backwards control-flow"
-        # a loop edge can't cross the boundary of a finalizer
-        b
+      ifs: seq[InstrPos]
+        ## stack of instruction positions for the currently open
+        ## structured control-flow blocks (if, loop, and regions).
+      blocks: seq[tuple[label: LabelId, hidden: bool]]
+        ## stack of targets for forward, mergin control-flow. `label` and
+        ## `hidden` are mutually exclusive.
```

Comment:
The default value for `LabelId` is a valid label, so that wouldn't work.

However, since `std/options` is now already imported, I think I'll change the tuple into an `Option`.